[Xcode] Dependencies in preprocessed sandbox profiles not tracked

emw-apple opened the pull request:
#### 40fd93b1f2eff242effa1c79d30571d1547c9a74
<pre>
[Xcode] Dependencies in preprocessed sandbox profiles not tracked
<a href="https://bugs.webkit.org/show_bug.cgi?id=257276">https://bugs.webkit.org/show_bug.cgi?id=257276</a>
rdar://problem/109790387

Reviewed by Brent Fulgham and Alexey Proskuryakov.

Reimplement the *.sb.in preprocessing task as a build rule, and have
clang emit a discovered dependency file. This means that the sandbox
profiles can be regenerated without effectively invalidating the entire
WebKit build, and that changes to common.sb and other included files
will cause profiles to regenerate.

Replicate this change in the iOS production-only &quot;Sandbox Profiles&quot;
target: First, change the target into a native static library so that
build rules are available, then add the *.sb.in profiles as resources.
(It doesn&apos;t have any sources, so it does not try to compile a static
library.)

* Source/WebKit/Configurations/WebKit.xcconfig: The project now tracks
  *.sb.in files as inputs, so update the exclude rule here.
* Source/WebKit/DerivedSources-input.xcfilelist: Remove *.sb.in entries.
* Source/WebKit/DerivedSources-output.xcfilelist: Remove *.sb entries.
* Source/WebKit/DerivedSources.make: Delete sandbox generator rule.
* Source/WebKit/Scripts/preprocess-sandbox-profile-rule.sh: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add build rules,
  target membership for *.sb.in files, and aforementioned Sandbox
  Profiles refactoring.

Canonical link: <a href="https://commits.webkit.org/264612@main">https://commits.webkit.org/264612@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fe450116acb5a8714255c52058bc01f38ff5a35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9756 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10374 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8298 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11053 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8245 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9324 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9877 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/8232 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6625 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14984 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7747 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/7542 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10892 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8009 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/7318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1952 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/11526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7752 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham reviewed (approved):
These changes look good to me, assuming EWS is happy.

aproskuryakov commented on Source/WebKit/Scripts/preprocess-sandbox-profile-rule.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh -e
+
+eval ARCHS=(${ARCHS})
```
Out of curiosity, what does this do?

emw-apple replied:
It turns ARCHS into an array, so that later we can grab the first architecture with `${ARCHS[0]}`. This is analogous to what the Makefiles does, `$(word 1, $(ARCHS))`. And it's needed because we have to run the clang preprocessor against _some_ target triple, and Copy Files phases are executed once per build, not once per architecture.

aproskuryakov commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -2238,6 +2238,21 @@
+			);
+			isEditable = 1;
+			outputFiles = (
+				"$(BUILT_PRODUCTS_DIR)/DerivedSources/WebKit/$(INPUT_FILE_BASE)",
```
`INPUT_FILE_BASE` removes `.in`, but leaves `.sb`?

emw-apple replied:
Exactly. It's the file name minus the last extension.

aproskuryakov commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -231,7 +231,7 @@ WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR = $(WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR_
 WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR_cocoatouch = $(PRIVATE_HEADERS_FOLDER_PATH);
 
-EXCLUDED_MACOS_PLUGIN_FILE_NAMES[sdk=iphone*] = SecItemShim.dylib WebProcessShim.dylib *.pdf Resources/mac/* com.apple.WebKit.NetworkProcess.sb com.apple.WebKit.GPUProcess.sb com.apple.WebKit.webpushd.mac.sb com.apple.WebProcess.sb;
+EXCLUDED_MACOS_PLUGIN_FILE_NAMES[sdk=iphone*] = SecItemShim.dylib WebProcessShim.dylib *.pdf Resources/mac/* com.apple.WebKit.NetworkProcess.sb.in com.apple.WebKit.GPUProcess.sb.in com.apple.WebKit.webpushd.mac.sb.in com.apple.WebProcess.sb.in;
```
I think the shims are gone, so you could remove them as you're touching this line.

aproskuryakov commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -211,11 +211,12 @@ OTHER_TAPI_FLAGS_SRCROOT_YES = -Xparser -include -Xparser config.h;
 EXCLUDED_PRODUCT_DEPENDENCY_NAMES = libWebKitAdditions.a AuthenticationServicesCore.framework;
 EXCLUDED_IOS_RESOURCE_FILE_NAMES = Resources/ios/*;
 EXCLUDED_IOS_RESOURCE_FILE_NAMES[sdk=iphone*] = ;
+EXCLUDED_MACOS_SOURCE_FILE_NAMES[sdk=iphone*] = Resources/mac/* com.apple.WebKit.NetworkProcess.sb.in com.apple.WebKit.GPUProcess.sb.in com.apple.WebKit.webpushd.mac.sb.in com.apple.WebProcess.sb.in;
```
Looks like `Resources/mac` is gone too, but I'm less certain about losing `*.pdf` from here.

emw-apple replied:
Good catch. I just checked, and there are no *.pdf files tracked in WebKit.xcodeproj. The exclude rule dates back to 2013, and unfortunately doesn't have a much context: d41656c0f04e00b3f3a48c0887282af31a442f0e

There are some PDFs inside of `iOS.xcassets`, but they are not known to the build system (since it processes the asset catalog as a unit).

emw-apple committed:
[Xcode] Dependencies in preprocessed sandbox profiles not tracked
https://bugs.webkit.org/show_bug.cgi?id=257276
rdar://problem/109790387

Reviewed by Brent Fulgham and Alexey Proskuryakov.

Reimplement the *.sb.in preprocessing task as a build rule, and have
clang emit a discovered dependency file. This means that the sandbox
profiles can be regenerated without effectively invalidating the entire
WebKit build, and that changes to common.sb and other included files
will cause profiles to regenerate.

Replicate this change in the iOS production-only "Sandbox Profiles"
target: First, change the target into a native static library so that
build rules are available, then add the *.sb.in profiles as resources.
(It doesn't have any sources, so it does not try to compile a static
library.)

* Source/WebKit/Configurations/WebKit.xcconfig: The project now tracks
  *.sb.in files as inputs, so update the exclude rule here.
* Source/WebKit/DerivedSources-input.xcfilelist: Remove *.sb.in entries.
* Source/WebKit/DerivedSources-output.xcfilelist: Remove *.sb entries.
* Source/WebKit/DerivedSources.make: Delete sandbox generator rule.
* Source/WebKit/Scripts/preprocess-sandbox-profile-rule.sh: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add build rules,
  target membership for *.sb.in files, and aforementioned Sandbox
  Profiles refactoring.

Canonical link: https://commits.webkit.org/264612@main
